// youtube-ext/dist/videoInfo.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.videoInfo = void 0;
const undici_1 = require("undici");
const constants_1 = require("./utils/constants");
const common_1 = require("./utils/common");
const extractStreamInfo_1 = require("./extractStreamInfo");
/**
 * Get full information about a YouTube video.
 */
const videoInfo = (url, options = {}) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, _51, _52, _53, _54, _55, _56, _57, _58, _59, _60, _61, _62, _63, _64, _65, _66;
    if (typeof url !== "string") {
        throw new Error(constants_1.constants.errors.type("url", "string", typeof url));
    }
    if (typeof options !== "object") {
        throw new Error(constants_1.constants.errors.type("options", "object", typeof options));
    }
    options = (0, common_1.mergeObj)({
        requestOptions: {
            headers: {
                "User-Agent": constants_1.constants.headers.userAgent,
            },
        },
    }, options);
    if (!url.startsWith("http")) {
        url = constants_1.constants.urls.video.base(url);
    }
    let data;
    try {
        const resp = yield (0, undici_1.request)(url, options.requestOptions);
        data = yield resp.body.text();
    }
    catch (err) {
        throw new Error(`Failed to fetch url "${url}". (${err})`);
    }
    let initialData;
    try {
        const initialDataRaw = (0, common_1.contentBetween)(data, "var ytInitialData = ", ";</script>");
        initialData = JSON.parse(initialDataRaw);
    }
    catch (err) {
        throw new Error(`Failed to parse data from webpage. (${err})`);
    }
    let initialPlayer;
    try {
        const initialPlayerRaw = (0, common_1.contentBetween)(data, "var ytInitialPlayerResponse = ", ";var meta = ");
        initialPlayer = JSON.parse(initialPlayerRaw);
    }
    catch (err) {
        throw new Error(`Failed to parse player data from webpage. (${err})`);
    }
    let contents;
    try {
        contents =
            (_d = (_c = (_b = (_a = initialData === null || initialData === void 0 ? void 0 : initialData.contents) === null || _a === void 0 ? void 0 : _a.twoColumnWatchNextResults) === null || _b === void 0 ? void 0 : _b.results) === null || _c === void 0 ? void 0 : _c.results) === null || _d === void 0 ? void 0 : _d.contents;
    }
    catch (err) {
        throw new Error(`Failed to parse contents from webpage. (${err})`);
    }
    let primary;
    try {
        primary = (_e = contents === null || contents === void 0 ? void 0 : contents.find((x) => x === null || x === void 0 ? void 0 : x.videoPrimaryInfoRenderer)) === null || _e === void 0 ? void 0 : _e.videoPrimaryInfoRenderer;
    }
    catch (err) { }
    let secondary;
    try {
        secondary = (_f = contents === null || contents === void 0 ? void 0 : contents.find((x) => x === null || x === void 0 ? void 0 : x.videoSecondaryInfoRenderer)) === null || _f === void 0 ? void 0 : _f.videoSecondaryInfoRenderer;
    }
    catch (err) { }
    const info = {
        title: (_h = (_g = primary === null || primary === void 0 ? void 0 : primary.title) === null || _g === void 0 ? void 0 : _g.runs[0]) === null || _h === void 0 ? void 0 : _h.text,
        id: (_k = (_j = initialData === null || initialData === void 0 ? void 0 : initialData.currentVideoEndpoint) === null || _j === void 0 ? void 0 : _j.watchEndpoint) === null || _k === void 0 ? void 0 : _k.videoId,
        url: constants_1.constants.urls.base +
            ((_o = (_m = (_l = initialData === null || initialData === void 0 ? void 0 : initialData.currentVideoEndpoint) === null || _l === void 0 ? void 0 : _l.commandMetadata) === null || _m === void 0 ? void 0 : _m.webCommandMetadata) === null || _o === void 0 ? void 0 : _o.url),
        shortDescription: (_p = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.videoDetails) === null || _p === void 0 ? void 0 : _p.shortDescription,
        description: (_s = (_r = (_q = secondary === null || secondary === void 0 ? void 0 : secondary.description) === null || _q === void 0 ? void 0 : _q.runs) === null || _r === void 0 ? void 0 : _r.map((x) => x === null || x === void 0 ? void 0 : x.text)) === null || _s === void 0 ? void 0 : _s.join(""),
        channel: {
            name: (_w = (_v = (_u = (_t = secondary === null || secondary === void 0 ? void 0 : secondary.owner) === null || _t === void 0 ? void 0 : _t.videoOwnerRenderer) === null || _u === void 0 ? void 0 : _u.title) === null || _v === void 0 ? void 0 : _v.runs[0]) === null || _w === void 0 ? void 0 : _w.text,
            id: (_2 = (_1 = (_0 = (_z = (_y = (_x = secondary === null || secondary === void 0 ? void 0 : secondary.owner) === null || _x === void 0 ? void 0 : _x.videoOwnerRenderer) === null || _y === void 0 ? void 0 : _y.title) === null || _z === void 0 ? void 0 : _z.runs[0]) === null || _0 === void 0 ? void 0 : _0.navigationEndpoint) === null || _1 === void 0 ? void 0 : _1.browseEndpoint) === null || _2 === void 0 ? void 0 : _2.browseId,
            url: constants_1.constants.urls.base +
                ((_8 = (_7 = (_6 = (_5 = (_4 = (_3 = secondary === null || secondary === void 0 ? void 0 : secondary.owner) === null || _3 === void 0 ? void 0 : _3.videoOwnerRenderer) === null || _4 === void 0 ? void 0 : _4.title) === null || _5 === void 0 ? void 0 : _5.runs[0]) === null || _6 === void 0 ? void 0 : _6.navigationEndpoint) === null || _7 === void 0 ? void 0 : _7.browseEndpoint) === null || _8 === void 0 ? void 0 : _8.canonicalBaseUrl),
            subscribers: {
                pretty: (_11 = (_10 = (_9 = secondary === null || secondary === void 0 ? void 0 : secondary.owner) === null || _9 === void 0 ? void 0 : _9.videoOwnerRenderer) === null || _10 === void 0 ? void 0 : _10.subscriberCountText) === null || _11 === void 0 ? void 0 : _11.simpleText,
            },
            icons: (_14 = (_13 = (_12 = secondary === null || secondary === void 0 ? void 0 : secondary.owner) === null || _12 === void 0 ? void 0 : _12.videoOwnerRenderer) === null || _13 === void 0 ? void 0 : _13.thumbnail) === null || _14 === void 0 ? void 0 : _14.thumbnails,
        },
        duration: {
            lengthSec: (_15 = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.videoDetails) === null || _15 === void 0 ? void 0 : _15.lengthSeconds,
        },
        thumbnails: (_17 = (_16 = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.videoDetails) === null || _16 === void 0 ? void 0 : _16.thumbnail) === null || _17 === void 0 ? void 0 : _17.thumbnails,
        ratings: {
            likes: {
                text: (_25 = (_24 = (_23 = (_22 = (_21 = (_20 = (_19 = (_18 = primary === null || primary === void 0 ? void 0 : primary.videoActions) === null || _18 === void 0 ? void 0 : _18.menuRenderer) === null || _19 === void 0 ? void 0 : _19.topLevelButtons) === null || _20 === void 0 ? void 0 : _20.find((x) => {
                    var _a, _b;
                    return ((_b = (_a = x === null || x === void 0 ? void 0 : x.toggleButtonRenderer) === null || _a === void 0 ? void 0 : _a.defaultIcon) === null || _b === void 0 ? void 0 : _b.iconType) ===
                        "LIKE";
                })) === null || _21 === void 0 ? void 0 : _21.toggleButtonRenderer) === null || _22 === void 0 ? void 0 : _22.defaultText) === null || _23 === void 0 ? void 0 : _23.accessibility) === null || _24 === void 0 ? void 0 : _24.accessibilityData) === null || _25 === void 0 ? void 0 : _25.label,
                pretty: (_31 = (_30 = (_29 = (_28 = (_27 = (_26 = primary === null || primary === void 0 ? void 0 : primary.videoActions) === null || _26 === void 0 ? void 0 : _26.menuRenderer) === null || _27 === void 0 ? void 0 : _27.topLevelButtons) === null || _28 === void 0 ? void 0 : _28.find((x) => {
                    var _a, _b;
                    return ((_b = (_a = x === null || x === void 0 ? void 0 : x.toggleButtonRenderer) === null || _a === void 0 ? void 0 : _a.defaultIcon) === null || _b === void 0 ? void 0 : _b.iconType) ===
                        "LIKE";
                })) === null || _29 === void 0 ? void 0 : _29.toggleButtonRenderer) === null || _30 === void 0 ? void 0 : _30.defaultText) === null || _31 === void 0 ? void 0 : _31.simpleText,
            },
            dislikes: {
                text: (_39 = (_38 = (_37 = (_36 = (_35 = (_34 = (_33 = (_32 = primary === null || primary === void 0 ? void 0 : primary.videoActions) === null || _32 === void 0 ? void 0 : _32.menuRenderer) === null || _33 === void 0 ? void 0 : _33.topLevelButtons) === null || _34 === void 0 ? void 0 : _34.find((x) => {
                    var _a, _b;
                    return ((_b = (_a = x === null || x === void 0 ? void 0 : x.toggleButtonRenderer) === null || _a === void 0 ? void 0 : _a.defaultIcon) === null || _b === void 0 ? void 0 : _b.iconType) ===
                        "DISLIKE";
                })) === null || _35 === void 0 ? void 0 : _35.toggleButtonRenderer) === null || _36 === void 0 ? void 0 : _36.defaultText) === null || _37 === void 0 ? void 0 : _37.accessibility) === null || _38 === void 0 ? void 0 : _38.accessibilityData) === null || _39 === void 0 ? void 0 : _39.label,
                pretty: (_45 = (_44 = (_43 = (_42 = (_41 = (_40 = primary === null || primary === void 0 ? void 0 : primary.videoActions) === null || _40 === void 0 ? void 0 : _40.menuRenderer) === null || _41 === void 0 ? void 0 : _41.topLevelButtons) === null || _42 === void 0 ? void 0 : _42.find((x) => {
                    var _a, _b;
                    return ((_b = (_a = x === null || x === void 0 ? void 0 : x.toggleButtonRenderer) === null || _a === void 0 ? void 0 : _a.defaultIcon) === null || _b === void 0 ? void 0 : _b.iconType) ===
                        "DISLIKE";
                })) === null || _43 === void 0 ? void 0 : _43.toggleButtonRenderer) === null || _44 === void 0 ? void 0 : _44.defaultText) === null || _45 === void 0 ? void 0 : _45.simpleText,
            },
        },
        views: {
            text: (_48 = (_47 = (_46 = primary === null || primary === void 0 ? void 0 : primary.viewCount) === null || _46 === void 0 ? void 0 : _46.videoViewCountRenderer) === null || _47 === void 0 ? void 0 : _47.viewCount) === null || _48 === void 0 ? void 0 : _48.simpleText,
            pretty: (_51 = (_50 = (_49 = primary === null || primary === void 0 ? void 0 : primary.viewCount) === null || _49 === void 0 ? void 0 : _49.videoViewCountRenderer) === null || _50 === void 0 ? void 0 : _50.shortViewCount) === null || _51 === void 0 ? void 0 : _51.simpleText,
        },
        published: {
            pretty: (_52 = primary === null || primary === void 0 ? void 0 : primary.dateText) === null || _52 === void 0 ? void 0 : _52.simpleText,
            text: (_54 = (_53 = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.microformat) === null || _53 === void 0 ? void 0 : _53.playerMicroformatRenderer) === null || _54 === void 0 ? void 0 : _54.publishDate,
        },
        uploaded: {
            text: (_56 = (_55 = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.microformat) === null || _55 === void 0 ? void 0 : _55.playerMicroformatRenderer) === null || _56 === void 0 ? void 0 : _56.uploadDate,
        },
        keywords: (_57 = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.videoDetails) === null || _57 === void 0 ? void 0 : _57.keywords,
        isLive: (_58 = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.videoDetails) === null || _58 === void 0 ? void 0 : _58.isLiveContent,
        isUnlisted: (_60 = (_59 = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.microformat) === null || _59 === void 0 ? void 0 : _59.playerMicroformatRenderer) === null || _60 === void 0 ? void 0 : _60.isUnlisted,
        isFamilySafe: (_62 = (_61 = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.microformat) === null || _61 === void 0 ? void 0 : _61.playerMicroformatRenderer) === null || _62 === void 0 ? void 0 : _62.isFamilySafe,
        category: (_64 = (_63 = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.microformat) === null || _63 === void 0 ? void 0 : _63.playerMicroformatRenderer) === null || _64 === void 0 ? void 0 : _64.category,
        embed: (_66 = (_65 = initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.microformat) === null || _65 === void 0 ? void 0 : _65.playerMicroformatRenderer) === null || _66 === void 0 ? void 0 : _66.embed,
        stream: initialPlayer === null || initialPlayer === void 0 ? void 0 : initialPlayer.streamingData,
    };
    (0, extractStreamInfo_1.prepareStreamInfo)(data, info.stream);
    return info;
});
exports.videoInfo = videoInfo;
exports.default = exports.videoInfo;
